#Illumina short read processing

All illumina paired-end RNA-seq data sets were downloaded from [SRA](https://www.ncbi.nlm.nih.gov/sra). To verify that there were no substantial library quality issues, We performed an initial qc with [fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) version 0.11.5. 

###Step1: kmer-based error correction
To correct sequencing errors independent of quality scores, we kmer-bsed error correction was performed with [rCorrector](https://github.com/mourisl/Rcorrector). For datasets consisting of multiple SRA accessions, accessions were corrected together. A generic 2-sample version of our command line is as follows:

    /path/to/rcorrector/run_rcorrector.pl -t 16 -1 sample1_R1.fq,sample2_R2.fq -2 sample1_R2.fq,sample2_R2.fq

###Step2: removing "unfixable" reads
In the headers of fastq files output by rCorrector, there is a flag indicating whether a read was corrected, or whether a read was deemed to contain an error, but that error was not fixable. We chose to remove read pairs where either read of a pair was flagged as unfixable. To do so, we ran the python script [FilterUncorrectablePEfastq.py](https://github.com/harvardinformatics/TranscriptomeAssemblyTools) available at the Harvard Informatics TranscriptomeAssemblyTools github repository. This script takes as input left and right reads, as well as a prefix to append to the output fastq files, e.g.
    
    Filter UncorrectablePEfastq.py -1 sample1_R1.cor.fastq -2 sample1_R2.cor.fastq -o rmunfixable
 
This script outputs fastq files with shortened headers and appends "/1" and "/2" to the R1 and R2 files, and truncates the third line of every read to "+". Header shortening is done to avoid downstream errors generated by Trinity during the step where it creates fasta files using fastool. 

###Step3: adapter and quality trimming
To the unfixable-removed paired-end reads, we applied [Trim Galore!](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/) to remove remaining adapter nucleotides and bases with a quality score less than 5. If TrimGalore!is not supplied an adapter nucleotide sequence, it uses the sequences for the standard illumina paired-end adapters. We supplied custom adapter sequences when we knew beforehand that those sequences were different from the standard illumina adapters, or when final analysis with fastqc indicated residual adapter contamination. An example command line is as follows:

    /path/to/TrimGalore!/trim_galore --paired --retain_unpaired --phred33 --output_dir `pwd` --length 36 -q 5 --stringency 5 -e 0.1 sample1_R1.fq sample1_R2.fq 

      
