# Intra-assembly redundancy

To assess the redundancy within assemblies across all assembled contigs, we perform a BLAT all-by-all alignment. We used a similar command line to that used to align contigs to reference transcripts for the analysis of transcript coverage, i.e.

    blat assembly.fasta assembly.fasta -t=dna -q=dna -fastMap allbyall.psl

with the only difference that we partitioned the assembly fasta into pieces for alignment to the whole transcriptome so that we could parallelize alignment across a job array.

Because in an all-by-all analysis, every pair of aligned contigs will show up twice (one will be a query the other a target, and vice versa), we removed duplicate entires with an rscript,[step1_BlatDedup.R](https://github.com/harvardinformatics/TranscriptomeAssemblyEvaluation/blob/master/assembly_redundancy/contig_redundancy/step1_BlatDedup.R). We then summarize redundancy between contigs using [Step2_SummarizeDeDupBlatAllByAllAlignments.py](https://github.com/harvardinformatics/TranscriptomeAssemblyEvaluation/blob/master/assembly_redundancy/contig_redundancy/step2_SummarizeDeDupBlatAllByAllAlignments.py), which takes as input the de-duplicated psl all-by-all alignment, an outfile name, a minimum contig size, a minimum alignment proportion, and a flag indicating whether the assembly is generated by Trinity. If it is a Trinity assembly, this flag uses the "gene"-"isoform" (i.e. component-contig) relationships to classify hits as intra or inter-component. 
